### PR TITLE
futhark: init at 0.6.2

### PR DESCRIPTION
**Pull request by infinisil**
###### Motivation for this change
Predicate to be able to demonstrate how well Nix can work as Futhark's package manager for https://www.reddit.com/r/ProgrammingLanguages/comments/90fnop/the_future_futhark_package_manager/

Edit: It's live :) https://github.com/infinisil/futharknixdemo

Tested with a simple compilation and run. It needed some fancy gcc `NIX_` stuff to work, @cleverca22 helped a lot with that on IRC :).

It's a Haskell package not present in hackage. I used cabal2nix to generate a file with the dependencies and made the updating process automated via a small script.

Ping @peti @domenkozar for Haskell

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/compilers/futhark/package.nix [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+, process-extras, QuickCheck, random, raw-strings-qq, regex-tdfa
+, srcloc, stdenv, template-haskell, temporary, test-framework
+, test-framework-hunit, test-framework-quickcheck2, text
+, th-lift-instances, transformers, vector, vector-binary-instances
+, zlib
+}:
+mkDerivation {
+  pname = "futhark";
+  version = "0.6.2";
+  src = /nix/store/5n2kpq21b3hm0yfr7wsinh5v24y81s98-source;
```
Comment (Infinisil) [inline pkgs/development/compilers/futhark/package.nix]: While a bit ugly, this is always overridden by the `default.nix`.

Comment (mpickering) [reply pkgs/development/compilers/futhark/package.nix]: Why not take the `src` as an argument? This is too ugly.

File: pkgs/development/compilers/futhark/default.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+let
+  cabal2nixPackage = callPackage ./package.nix {};
+  src = callPackage ./src.nix {};
+
+  package = cabal2nixPackage.overrideAttrs (drv: { inherit src; });
+
+  path = stdenv.lib.makeBinPath [ gcc ];
+
+  wrapped = symlinkJoin {
```
Comment (mpickering) [inline pkgs/development/compilers/futhark/default.nix]: Why do you use `symlinkJoin` with one one thing in `paths`?

File: pkgs/development/compilers/futhark/update [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+expr="(import ../../../.. {}).callPackage ./src.nix {}"
+src=$(nix-build --no-out-link -E "$expr")
+if cabal2nix "$src" --hpack > tmp.nix; then
+	mv tmp.nix package.nix
+else
+	rm tmp.nix
+	exit 1
+fi
+
+echo Updated package.nix file successfully
```
Comment (mpickering) [inline pkgs/development/compilers/futhark/update]: I feel like if you have to use this script then it should be packaged properly and automatically run on whatever `src` you choose rather than just living in the directory.

File: pkgs/development/compilers/futhark/default.nix (hunk 2 above)
Comment (Infinisil) [reply pkgs/development/compilers/futhark/default.nix]: Makes it easy to patch a single file, since everything gets symlinked and I can just replace a symlink. Idris also uses this in its wrapper.

Alternative would be to do `ln -s ${package}/* $out; rm $out/bin; mkdir $out/bin; ln -s ${package}/bin/* $out/bin; rm $out/bin/futhark-c; rm $out/bin/futhark-opencl` and then the makeWrappers. I can do that if you prefer.

File: pkgs/development/compilers/futhark/package.nix (hunk 1 above)
Comment (Infinisil) [reply pkgs/development/compilers/futhark/package.nix]: This is autogenerated by cabal2nix, I can't just change what it generates. I also don't want to have to manually edit the file every time an update comes out. It would be possible to use `cabal2nix`'s git fetching. But that has the problem that it doesn't use the github tarballs, it fetches all submodules, and futhark has a super big submodule that takes forever to fetch. This may get better with https://github.com/NixOS/cabal2nix/issues/260. I don't know of a cleaner solution than this.

**Comment by Infinisil** [comment]
@mpickering I removed the update script and replaced it with a short manual description of how to update in the file, is this alright?

**Comment by mpickering** [comment]
Where is the description?

**Comment by mpickering** [comment]
OK, I think this is fine. It seems annoying that there isn't one invocation of `cabal2nix` which will generate this file but I can't get it to work.

For example:

```
cabal2nix https://github.com/diku-dk/futhark --hpack --revision v0.6.2
```

**Comment by Infinisil** [comment]
Yeah currently there isn't, as previously mentioned, there is an issue for it: https://github.com/NixOS/cabal2nix/issues/260

**Comment by Infinisil** [comment]
@peti @basvandijk Is it okay to add a haskell package like this? It's not in hackage

**Comment by flokli** [comment]
Was able to compile some examples with it. Unfortunately couldn't execute them, as I couldn't get OpenCL to work on my hardware (yet).

`opencl-info` complains
```
ERROR: clGetPlatformIDs(-1001)
```

Thinkpad T460p with i7-6820HQ, currently not using the nvidia card `NVIDIA Corporation GM108M [GeForce 940MX]` at all. According to [this](https://en.wikichip.org/wiki/intel/core_i7/i7-6820hq), the Intel Graphics should support OpenCL on it's own, so not really sure what do do here to get it working `¯\_(ツ)_/¯`

**Review by basvandijk — CHANGES_REQUESTED** [review]
@Infinisil cool compiler!

The fact that futhark is a Haskell package is an implementation detail. To install it users should just have to specify `futhark` instead of `haskellPackages.futhark`.

Note that we also do this for most other Haskell executables. Like those other Haskell executables we should also override futhark with `haskell.lib.justStaticExecutables` to
reduce closure size.

I made these changes in: https://github.com/LumiGuide/nixpkgs/commit/5ba57d552090eeb029ff2bfd6ddd707542420902. You may want to merge that into your branch or I can also do it if you like.

Offtopic: note that in https://github.com/NixOS/nixpkgs/pull/45406 I'm disabling library profiling for `justStaticExecutables` meaning that once that is merged `futhark` will build much quicker.

File: pkgs/development/haskell-modules/configuration-nix.nix [hunk 4]
```diff
@@ -510,4 +510,20 @@ self: super: builtins.intersectAttrs super {
...
   }));
+
+  futhark = with pkgs;
+    let path = stdenv.lib.makeBinPath [ gcc ];
+    in overrideCabal (addBuildTool super.futhark makeWrapper) (_drv: {
+      postInstall = ''
+        wrapProgram $out/bin/futhark-c \
+          --prefix PATH : "${path}"
+
+        wrapProgram $out/bin/futhark-opencl \
```
Comment (Infinisil) [inline pkgs/development/haskell-modules/configuration-nix.nix]: @basvandijk These wrapper don't seem to be applied like this, building either `pkgs.futhark` or `pkgs.haskellPackages.futhark` results in raw ELF binaries

Comment (basvandijk) [reply pkgs/development/haskell-modules/configuration-nix.nix]: Strange, I did test if it created the correct wrappers but I can confirm it now results in ELF binaries. I'll look into it.

Comment (basvandijk) [reply pkgs/development/haskell-modules/configuration-nix.nix]: This should now be fixed by https://github.com/NixOS/nixpkgs/pull/43865/commits/2cdeee2367ccc478b25a6e1db52d099e3d258896 which I pushed to this PR.

@peti OK with this solution?

**Review by basvandijk — APPROVED** [review]

**Comment by basvandijk** [comment]
@GrahamcOfBorg build futhark

**Comment by basvandijk** [comment]
@GrahamcOfBorg build haskell-ci
since I also changed that in this PR.

@Infinisil I'll let you do the honours of merging this in if you think it's OK now.

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -8423,7 +8425,7 @@ with pkgs;
 
   msitools = callPackage ../development/tools/misc/msitools { };
 
-  multi-ghc-travis = haskell.lib.justStaticExecutables haskellPackages.multi-ghc-travis;
+  haskell-ci = haskell.lib.justStaticExecutables haskellPackages.haskell-ci;
```
Comment (Infinisil) [inline pkgs/top-level/all-packages.nix]: @basvandijk Why this change?

Comment (basvandijk) [reply pkgs/top-level/all-packages.nix]: While moving the non-hackage packages to a dedicated file I discovered that `multi-ghc-travis` has been renamed to `haskell-ci` and that `haskell-ci` is now on Hackage. This is explained in: https://github.com/NixOS/nixpkgs/commit/6106664ffab6dd5f6f2c1191969545d6b4f88ee1

Comment (basvandijk) [reply pkgs/top-level/all-packages.nix]: Mmm maybe an alias might be in order here...

**Comment by Infinisil** [comment]
Rebased to fix conflicts, tested the binary again (still works) and added a `multi-ghc-travis` with an explanatory error message.
